include/deal.II/base/symbolic_function.h

```diff
@@ -0,0 +1,470 @@
+
+    /**
+     * Store the user substitution map used for expression substitutions. This
+     * may be updated with a call to update_user_substitution_map(). Notice that
+     * the function may still have unresolved symbols, provided that they are
+     * resolved by a call to set_additional_function_arguments().
+     */
+    Differentiation::SD::types::substitution_map user_substitution_map;
+
+    /**
+     * Store a user substitution map used for additional argument
+     * substitutions. This will be updated by a call to
+     * set_additional_function_arguments().
+     */
+    Differentiation::SD::types::substitution_map additional_function_arguments;
+
+    /**
+     * The actual components of this symbolic function. This is obtained from
+     * the @p user_function, after applying the @p user_substitution_map.
+     */
+    mutable std::vector<Differentiation::SD::Expression> function;
+
+    /**
+     * The gradients of each component of this symbolic function. This is
+     * obtained by computing the symbolic gradient of the object @p function,
+     * that is, after applying the @p user_substitution_map to @p user_function.
+     */
+    mutable std::vector<Tensor<1, dim, Differentiation::SD::Expression>>
+      function_gradient;
+
+    /**
+     * The Hessians of each component of this symbolic function. This is
+     * obtained by computing the symbolic Hessian of the object @p function,
+     * that is, after applying the @p user_substitution_map to @p user_function.
+     */
+    mutable std::vector<Tensor<2, dim, Differentiation::SD::Expression>>
+      function_hessian;
+
+    /**
+     * The Laplacians of each component of this symbolic function. This is
+     * obtained by computing the symbolic Laplacian of the object @p function,
+     * that is, after applying the @p user_substitution_map to @p user_function.
+     */
+    mutable std::vector<Differentiation::SD::Expression> function_laplacian;
+
+    /**
+     * The coordinate symbols argument of the function.
+     */
+    Tensor<1, dim, Differentiation::SD::Expression> coordinate_symbols;
+
+    /**
+     * The symbolic time argument of the function.
+     */
+    mutable Differentiation::SD::Expression time_symbol;
+  };
+
+  /**
+   * Allow output using the bitwise left shift operator.
+   */
+  template <int dim, typename RangeNumberType>
+  inline std::ostream &
+  operator<<(std::ostream &out, const SymbolicFunction<dim, RangeNumberType> &f)
+  {
+    return f.print(out);
+  }
+
+
+
+  // Inline and template functions
+  template <int dim, typename RangeNumberType>
+  template <typename StreamType>
+  StreamType &
+  SymbolicFunction<dim, RangeNumberType>::print(StreamType &out) const
+  {
+    for (unsigned int i = 0; i < dim; ++i)
+      out << coordinate_symbols[i] << ", ";
+    out << time_symbol << " -> " << user_function[0];
+    for (unsigned int i = 1; i < user_function.size(); ++i)
+      out << "; " << user_function[i];
```

Comment:
I guess that, in order to match the description of the `print` function, we need to revert this to printing the `function` itself. Alternatively, we could print the keys in the `user_substitution_map` and `additional_function_arguments` (if not empty), which I think might actually be nicer (add to the list of coord symbols and time symbol before the `->`).

Reply:
I've adopted a strategy which is mixed. Arguments are treated in the same way of the coordinates and of time, while the user_subsitution_map is appended after a `#` symbol, providing the full picture of the function.

Reply:
Let me know what you think of the output of test n. 4

Reply:
Looks great. Thank you :-)